src/Client.cpp

```diff
@@ -93,120 +71,105 @@ void UrsaClient::init_conn(zmq::socket_t &socket) {
     }
 }
 
-void UrsaClient::recv_res(zmq::socket_t &socket) {
-    // enable periodic progress check
-    // which is done by status worker
-    this->command_active = true;
-    this->wait_time = 0;
-
-    while (this->command_active) {
-        auto res_str = s_recv<std::string>(&socket);
-
-        if (res_str.empty()) {
-            continue;
+void UrsaClient::recv_res(zmq::socket_t *socket) {
+    std::string result;
+    while (result.empty()) {
+        try {
+            result = s_recv<std::string>(socket);
+        } catch (const std::runtime_error &ex) {
+            check_task_status(this->connection_id);
         }
+    }
 
-        this->command_active = false;
-        auto res = json::parse(res_str);
-
-        if (this->raw_json) {
-            std::cout << res.dump(4) << std::endl;
-            return;
-        }
+    auto res = json::parse(result);
+    if (this->raw_json) {
+        std::cout << res.dump(4) << std::endl;
+        return;
+    }
 
-        if (res["type"] == "select") {
-            for (const auto &file : res["result"]["files"]) {
-                std::cout << file.get<std::string>() << std::endl;
-            }
-        } else if (res["type"] == "error") {
-            spdlog::error(res["error"]["message"].get<std::string>());
-        } else {
-            std::cout << res.dump(4) << std::endl;
+    if (res["type"] == "select") {
+        for (const auto &file : res["result"]["files"]) {
+            std::cout << file.get<std::string>() << std::endl;
         }
+    } else if (res["type"] == "error") {
+        spdlog::error(res["error"]["message"].get<std::string>());
+    } else {
+        std::cout << res.dump(4) << std::endl;
     }
 }
 
-static void s_send_cmd(zmq::socket_t &socket, std::string cmd) {
+static void s_send_cmd(zmq::socket_t *socket, std::string cmd) {
     // for user convenience
     if (cmd.back() != ';') {
         cmd = cmd + ";";
     }
-
-    s_send(&socket, cmd);
+    s_send(socket, cmd);
 }
 
-int UrsaClient::start() {
-    zmq::context_t context(1);
-    zmq::socket_t socket(context, ZMQ_REQ);
-    socket.setsockopt(ZMQ_LINGER, 0);
-    socket.setsockopt(ZMQ_RCVTIMEO, 1000);
-    socket.connect(server_addr);
+void UrsaClient::setup_connection() {
```

Comment:
shouldn't this be done once in the constructor, instead of before one_shot_command and start()?

For example it'd make sense to use UrsaClient like this:

```
auto client = ...
client.one_shot_command("aaa");
client.one_shot_command("bbb");
client.one_shot_command("ccc");
```

(ok, this is not useful yet since one_shot will only print the command, but who knows in the future!)